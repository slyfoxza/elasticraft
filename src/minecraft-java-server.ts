/* Copyright 2025 Philip Cronje
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
import * as ec2 from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";

export interface MinecraftJavaServerProps {
  instanceType?: ec2.InstanceType;
}

export class MinecraftJavaServer extends Construct {
  readonly launchTemplate: ec2.ILaunchTemplate;

  constructor(scope: Construct, id: string, props?: MinecraftJavaServerProps) {
    super(scope, id);

    const instanceType =
      props?.instanceType ??
      ec2.InstanceType.of(ec2.InstanceClass.T4G, ec2.InstanceSize.MEDIUM);
    this.launchTemplate = new ec2.LaunchTemplate(this, "LaunchTemplate", {
      instanceType,
      machineImage: ec2.MachineImage.resolveSsmParameterAtLaunch(
        ec2.AmazonLinux2023ImageSsmParameter.ssmParameterName({
          cpuType:
            instanceType.architecture as unknown as ec2.AmazonLinuxCpuType,
        }),
      ),
    });
  }
}
