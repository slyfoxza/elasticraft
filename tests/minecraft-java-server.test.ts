/* Copyright 2025 Philip Cronje
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
import { rm } from "node:fs/promises";

import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { App, Stack } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import * as ec2 from "aws-cdk-lib/aws-ec2";

import { MinecraftJavaServer } from "../src/minecraft-java-server.js";

describe(MinecraftJavaServer, () => {
  let stack: Stack;

  beforeEach(() => {
    stack = new Stack();
  });

  afterEach(async () => {
    const outputDirectory = App.of(stack)?.outdir;
    if (outputDirectory !== undefined) {
      await rm(outputDirectory, { recursive: true });
    }
  });

  it("generates the expected template", () => {
    new MinecraftJavaServer(stack, "Test");
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("accepts a different instance type", () => {
    new MinecraftJavaServer(stack, "Test", {
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.C7I,
        ec2.InstanceSize.LARGE,
      ),
    });
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::EC2::LaunchTemplate", {
      LaunchTemplateData: { InstanceType: "c7i.large" },
    });
  });

  it("accepts a custom server ID", () => {
    new MinecraftJavaServer(stack, "Test", { serverId: "my-server" });
    const template = Template.fromStack(stack);
    const serverIdTag = { Key: "elasticraft:serverId", Value: "my-server" };
    template.hasResourceProperties("AWS::EC2::LaunchTemplate", {
      LaunchTemplateData: {
        TagSpecifications: [
          { ResourceType: "instance", Tags: Match.arrayWith([serverIdTag]) },
          { ResourceType: "volume", Tags: Match.arrayWith([serverIdTag]) },
        ],
      },
      TagSpecifications: [
        {
          ResourceType: "launch-template",
          Tags: Match.arrayWith([serverIdTag]),
        },
      ],
    });
  });
});
