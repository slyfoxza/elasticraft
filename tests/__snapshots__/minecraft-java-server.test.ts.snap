// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MinecraftJavaServer > generates the expected template 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestLaunchTemplate2212C1A1": {
      "Properties": {
        "LaunchTemplateData": {
          "ImageId": "resolve:ssm:/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-arm64",
          "InstanceInitiatedShutdownBehavior": "terminate",
          "InstanceType": "t4g.medium",
          "MetadataOptions": {
            "HttpTokens": "required",
          },
          "TagSpecifications": [
            {
              "ResourceType": "instance",
              "Tags": [
                {
                  "Key": "elasticraft:serverId",
                  "Value": "Test",
                },
                {
                  "Key": "Name",
                  "Value": "Default/Test/LaunchTemplate",
                },
              ],
            },
            {
              "ResourceType": "volume",
              "Tags": [
                {
                  "Key": "elasticraft:serverId",
                  "Value": "Test",
                },
                {
                  "Key": "Name",
                  "Value": "Default/Test/LaunchTemplate",
                },
              ],
            },
          ],
          "UserData": {
            "Fn::Base64": {
              "Fn::Join": [
                "",
                [
                  "Content-Type: multipart/mixed; boundary="+AWS+CDK+User+Data+Separator=="
MIME-Version: 1.0

--+AWS+CDK+User+Data+Separator==
Content-Type: text/x-shellscript; charset="utf-8"
Content-Transfer-Encoding: base64

",
                  {
                    "Fn::Base64": "#!/bin/bash
mkdir --parents /etc/elasticraft
echo -n 'Test' > /etc/elasticraft/server-id",
                  },
                  "
--+AWS+CDK+User+Data+Separator==--
",
                ],
              ],
            },
          },
        },
        "TagSpecifications": [
          {
            "ResourceType": "launch-template",
            "Tags": [
              {
                "Key": "elasticraft:serverId",
                "Value": "Test",
              },
              {
                "Key": "Name",
                "Value": "Default/Test/LaunchTemplate",
              },
            ],
          },
        ],
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
